infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPersistentVolumeClaims.java

```diff
@@ -137,4 +148,86 @@ public void delete(Map<String, String> labels) throws InfrastructureException {
       throw new KubernetesInfrastructureException(e);
     }
   }
+
+  /**
+   * Waits until persistent volume claim state will bound.
+   *
+   * @param name name of persistent volume claim that should be watched
+   * @param timeoutMillis waiting timeout in milliseconds
+   * @return persistent volume claim that satisfies the specified predicate
+   * @throws InfrastructureException when specified timeout is reached
+   * @throws InfrastructureException when {@link Thread} is interrupted while waiting
+   * @throws InfrastructureException when any other exception occurs
+   */
+  public PersistentVolumeClaim waitBound(String name, long timeoutMillis)
+      throws InfrastructureException {
+    return wait(name, timeoutMillis, pvc -> pvc.getStatus().getPhase().equals(PVC_BOUND_PHASE));
+  }
+
+  /**
+   * Waits until persistent volume claim state will suit for specified predicate.
+   *
+   * @param name name of persistent volume claim that should be watched
+   * @param timeoutMillis waiting timeout in milliseconds
+   * @param predicate predicate to perform state check
+   * @return persistent volume claim that satisfies the specified predicate
+   * @throws InfrastructureException when specified timeout is reached
+   * @throws InfrastructureException when {@link Thread} is interrupted while waiting
+   * @throws InfrastructureException when any other exception occurs
+   */
+  public PersistentVolumeClaim wait(
+      String name, long timeoutMillis, Predicate<PersistentVolumeClaim> predicate)
+      throws InfrastructureException {
+    CompletableFuture<PersistentVolumeClaim> future = new CompletableFuture<>();
+    Watch watch = null;
+    try {
+      Resource<PersistentVolumeClaim, DoneablePersistentVolumeClaim> pvcResource =
+          clientFactory
+              .create(workspaceId)
+              .persistentVolumeClaims()
+              .inNamespace(namespace)
+              .withName(name);
+
+      watch =
+          pvcResource.watch(
+              new Watcher<PersistentVolumeClaim>() {
+                @Override
+                public void eventReceived(Action action, PersistentVolumeClaim pvc) {
+                  if (predicate.test(pvc)) {
+                    future.complete(pvc);
+                  }
+                }
+
+                @Override
+                public void onClose(KubernetesClientException cause) {
+                  future.completeExceptionally(
+                      new InfrastructureException(
+                          "Waiting for persistent volume claim '" + name + "' was interrupted"));
+                }
+              });
+
+      PersistentVolumeClaim actualPvc = pvcResource.get();
+      if (predicate.test(actualPvc)) {
+        return actualPvc;
+      }
+      try {
+        return future.get(timeoutMillis, TimeUnit.MILLISECONDS);
+      } catch (ExecutionException e) {
+        throw new InfrastructureException(e.getCause().getMessage(), e);
```

Comment:
Please consider throwing here InternalInfrastructureException and add code comments to not forget the motivation why we choose some particular behavior